File: fabric/src/main/java/com/suprememc/SupremeMC.java
package com.suprememc;

import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.biome.v1.BiomeModifications;
import net.fabricmc.fabric.api.biome.v1.BiomeSelectors;
import net.fabricmc.fabric.api.loot.v3.LootTableEvents;
import net.fabricmc.fabric.api.registry.FuelValueEvents;
import net.fabricmc.fabric.api.registry.FabricPotionBrewingBuilder;
import net.fabricmc.fabric.api.object.builder.v1.entity.FabricDefaultAttributeRegistry;
import com.suprememc.content.ModContent;
import com.suprememc.loot.ModLootInjections;
import net.minecraft.core.Registry;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.Identifier;
import net.minecraft.resources.ResourceKey;
import net.minecraft.tags.BiomeTags;
import net.minecraft.world.level.biome.Biomes;
import net.minecraft.world.item.CreativeModeTab;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.alchemy.Potions;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.MobCategory;
import net.minecraft.world.level.block.entity.BlockEntityTypes;
import net.minecraft.world.level.levelgen.GenerationStep;
import net.minecraft.world.level.storage.loot.LootPool;
import net.minecraft.world.level.storage.loot.LootTable;
import net.minecraft.world.level.storage.loot.entries.NestedLootTable;
import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
import net.minecraft.world.level.biome.Biome;
import net.fabricmc.fabric.api.biome.v1.BiomeSelectionContext;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Predicate;

public class SupremeMC implements ModInitializer {

    private static final ResourceKey<Biome> CAYS =
        ResourceKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "cays"));

    @Override
    public void onInitialize() {

        // This method is invoked by the Fabric mod loader when it is ready
        // to load your mod. You can access Fabric and Common code in this
        // project.

        // Use Fabric to bootstrap the Common mod.
        Constants.LOG.info("Hello Fabric world!");
        CommonClass.init();
        FabricPotionBrewingBuilder.BUILD.register(builder -> {
            builder.registerPotionRecipe(net.minecraft.world.item.alchemy.Potions.AWKWARD,
                Ingredient.of(ModContent.CLOVER.asItem()), ModContent.LUCK_POTION);
            builder.registerPotionRecipe(ModContent.LUCK_POTION, Ingredient.of(Items.REDSTONE), ModContent.LONG_LUCK_POTION);
            builder.registerPotionRecipe(ModContent.LUCK_POTION, Ingredient.of(Items.GLOWSTONE_DUST), ModContent.STRONG_LUCK_POTION);
            builder.registerPotionRecipe(ModContent.LUCK_POTION, Ingredient.of(Items.FERMENTED_SPIDER_EYE), ModContent.BAD_LUCK_POTION);
            builder.registerPotionRecipe(ModContent.LONG_LUCK_POTION, Ingredient.of(Items.FERMENTED_SPIDER_EYE), ModContent.LONG_BAD_LUCK_POTION);
            builder.registerPotionRecipe(ModContent.STRONG_LUCK_POTION, Ingredient.of(Items.FERMENTED_SPIDER_EYE), ModContent.STRONG_BAD_LUCK_POTION);
                builder.registerPotionRecipe(Potions.AWKWARD, Ingredient.of(Items.ROTTEN_FLESH), ModContent.HUNGER_POTION);
                builder.registerPotionRecipe(ModContent.HUNGER_POTION, Ingredient.of(Items.REDSTONE), ModContent.LONG_HUNGER_POTION);
                builder.registerPotionRecipe(ModContent.HUNGER_POTION, Ingredient.of(Items.GLOWSTONE_DUST), ModContent.STRONG_HUNGER_POTION);
                builder.registerPotionRecipe(Potions.AWKWARD, Ingredient.of(Items.WITHER_ROSE), ModContent.DECAY_POTION);
                builder.registerPotionRecipe(ModContent.DECAY_POTION, Ingredient.of(Items.REDSTONE), ModContent.LONG_DECAY_POTION);
                builder.registerPotionRecipe(ModContent.DECAY_POTION, Ingredient.of(Items.GLOWSTONE_DUST), ModContent.STRONG_DECAY_POTION);
        });
        FuelValueEvents.BUILD.register((builder, context) -> {
            builder.add(ModContent.ANTHRACITE, 1600);
            builder.add(ModContent.ANTHRACITE_BLOCK, 16000);
        });
        FabricDefaultAttributeRegistry.register(ModContent.GRIZZLY_BEAR_ENTITY,
            net.minecraft.world.entity.animal.polarbear.PolarBear.createAttributes());
        FabricDefaultAttributeRegistry.register(ModContent.ENDER_SPIDER_ENTITY,
            com.suprememc.content.entity.EnderSpider.createAttributes());
        FabricDefaultAttributeRegistry.register(ModContent.FIRE_CREEPER_ENTITY,
            net.minecraft.world.entity.monster.Creeper.createAttributes());
        FabricDefaultAttributeRegistry.register(ModContent.SNOW_CREEPER_ENTITY,
            net.minecraft.world.entity.monster.Creeper.createAttributes());
        // Vanilla's sign block entities only accept their hardcoded block list, so opt the palm signs in.
        BlockEntityTypes.SIGN.addValidBlock(ModContent.PALM_SIGN);
        BlockEntityTypes.SIGN.addValidBlock(ModContent.PALM_WALL_SIGN);
        BlockEntityTypes.HANGING_SIGN.addValidBlock(ModContent.PALM_HANGING_SIGN);
        BlockEntityTypes.HANGING_SIGN.addValidBlock(ModContent.PALM_WALL_HANGING_SIGN);
        BlockEntityTypes.FURNACE.addValidBlock(ModContent.BLACKSTONE_FURNACE);
        BlockEntityTypes.FURNACE.addValidBlock(ModContent.DEEPSLATE_FURNACE);
        BiomeModifications.addFeature(
            BiomeSelectors.tag(BiomeTags.IS_OCEAN),
            GenerationStep.Decoration.UNDERGROUND_ORES,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "aquamarine_ore")));
        BiomeModifications.addFeature(
            BiomeSelectors.tag(BiomeTags.IS_OCEAN),
            GenerationStep.Decoration.UNDERGROUND_ORES,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "prismarine_ore")));
        BiomeModifications.addFeature(
            BiomeSelectors.tag(BiomeTags.IS_OCEAN),
            GenerationStep.Decoration.UNDERGROUND_ORES,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "prismarine_ore_small")));
        BiomeModifications.addFeature(
            BiomeSelectors.tag(BiomeTags.IS_OCEAN),
            GenerationStep.Decoration.UNDERGROUND_ORES,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "atlantis_debris")));
        BiomeModifications.addFeature(
            BiomeSelectors.includeByKey(Biomes.FOREST, Biomes.FLOWER_FOREST, Biomes.BIRCH_FOREST,
                Biomes.OLD_GROWTH_BIRCH_FOREST, Biomes.DARK_FOREST, Biomes.GROVE),
            GenerationStep.Decoration.UNDERGROUND_ORES,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "amber_ore")));
        BiomeModifications.addFeature(
            BiomeSelectors.includeByKey(Biomes.PALE_GARDEN),
            GenerationStep.Decoration.UNDERGROUND_ORES,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "amber_ore_pale_garden")));
        BiomeModifications.addFeature(
            BiomeSelectors.tag(BiomeTags.IS_NETHER),
            GenerationStep.Decoration.UNDERGROUND_ORES,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "nether_anthracite_ore")));
        BiomeModifications.addFeature(
            BiomeSelectors.tag(BiomeTags.IS_NETHER),
            GenerationStep.Decoration.UNDERGROUND_ORES,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "burning_diamond_ore")));
        // Palm beaches: warm-climate beaches always grow palms; moderate-climate beaches at a 50% chunk rate
        // and lower density. The placed feature JSONs (packaged from the shared generated data) carry the
        // temperature tiers and sand substrate rules, so behavior matches the NeoForge biome modifiers.
        BiomeModifications.addFeature(
            BiomeSelectors.tag(BiomeTags.IS_BEACH),
            GenerationStep.Decoration.VEGETAL_DECORATION,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "palm_trees_warm")));
        BiomeModifications.addFeature(
            BiomeSelectors.tag(BiomeTags.IS_BEACH),
            GenerationStep.Decoration.VEGETAL_DECORATION,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "palm_trees_temperate")));
        BiomeModifications.addFeature(
            BiomeSelectors.includeByKey(Biomes.PLAINS, Biomes.SUNFLOWER_PLAINS),
            GenerationStep.Decoration.VEGETAL_DECORATION,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "cotton_bushes")));
        BiomeModifications.addFeature(
            BiomeSelectors.includeByKey(Biomes.PLAINS, Biomes.SUNFLOWER_PLAINS),
            GenerationStep.Decoration.VEGETAL_DECORATION,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "corn_patches")));
        // Beach grass grows on both real beach biomes and the custom Cays biome (mirrors the NeoForge biome modifier).
        Predicate<BiomeSelectionContext> beachOrCays =
            BiomeSelectors.tag(BiomeTags.IS_BEACH).or(BiomeSelectors.includeByKey(CAYS));
        BiomeModifications.addFeature(
            beachOrCays,
            GenerationStep.Decoration.VEGETAL_DECORATION,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "beach_grass")));
        BiomeModifications.addFeature(
            beachOrCays,
            GenerationStep.Decoration.VEGETAL_DECORATION,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "tall_beach_grass")));
        BiomeModifications.addFeature(
            BiomeSelectors.tag(BiomeTags.IS_SAVANNA),
            GenerationStep.Decoration.VEGETAL_DECORATION,
            ResourceKey.create(Registries.PLACED_FEATURE, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "grape_vine_hang")));
        registerDrownedSpawns();
        registerEnderSpiderSpawns();
        registerFireCreeperSpawns();
        registerSnowCreeperSpawns();

        CreativeModeTab tab = CreativeModeTab.builder(CreativeModeTab.Row.TOP, 0)
                .title(Component.translatable("itemGroup." + Constants.MOD_ID + ".main"))
                .icon(() -> new ItemStack(ModContent.SUPREME_MC_LOGO_BLOCK))
                .displayItems((params, output) -> ModContent.CREATIVE_TAB_ITEMS.forEach(output::accept))
                .build();
        Registry.register(BuiltInRegistries.CREATIVE_MODE_TAB,
                ResourceKey.create(Registries.CREATIVE_MODE_TAB, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "main")), tab);

        registerLootInjections();
    }

        private static void registerDrownedSpawns() {
        Predicate<BiomeSelectionContext> river = BiomeSelectors.includeByKey(Biomes.RIVER);
        Predicate<BiomeSelectionContext> frozenRiver = BiomeSelectors.includeByKey(Biomes.FROZEN_RIVER);
        Predicate<BiomeSelectionContext> dripstoneCaves = BiomeSelectors.includeByKey(Biomes.DRIPSTONE_CAVES);
        Predicate<BiomeSelectionContext> iceCaves = BiomeSelectors.includeByKey(ResourceKey.create(
            Registries.BIOME, Identifier.fromNamespaceAndPath(Constants.MOD_ID, "ice_caves")));
        Predicate<BiomeSelectionContext> oceans = BiomeSelectors.tag(BiomeTags.IS_OCEAN);

        BiomeModifications.create(Identifier.fromNamespaceAndPath(Constants.MOD_ID, "drowned_spawns"))
            .add(net.fabricmc.fabric.api.biome.v1.ModificationPhase.REPLACEMENTS, river,
                context -> replaceDrownedSpawn(context, 100, 1, 1))
            .add(net.fabricmc.fabric.api.biome.v1.ModificationPhase.REPLACEMENTS, dripstoneCaves,
                context -> replaceDrownedSpawn(context, 100, 4, 4))
            .add(net.fabricmc.fabric.api.biome.v1.ModificationPhase.REPLACEMENTS, iceCaves,
                context -> replaceDrownedSpawn(context, 100, 4, 4))
            .add(net.fabricmc.fabric.api.biome.v1.ModificationPhase.REPLACEMENTS, oceans,
                context -> replaceDrownedSpawn(context, 100, 1, 1))
            .add(net.fabricmc.fabric.api.biome.v1.ModificationPhase.REPLACEMENTS, frozenRiver,
                context -> replaceDrownedSpawn(context, 5, 1, 1));
        }

        private static void replaceDrownedSpawn(net.fabricmc.fabric.api.biome.v1.BiomeModificationContext context,
                            int weight, int minCount, int maxCount) {
            EntityType<?> drowned = BuiltInRegistries.ENTITY_TYPE.get(
                Identifier.fromNamespaceAndPath("minecraft", "drowned")).orElseThrow().value();
            context.getMobSpawnSettings().removeSpawnsOfEntityType(drowned);
        context.getMobSpawnSettings().addSpawn(
            MobCategory.MONSTER,
                new net.minecraft.world.level.biome.MobSpawnSettings.SpawnerData(drowned, minCount, maxCount),
            weight);
        }

        private static void registerFireCreeperSpawns() {
        BiomeModifications.create(Identifier.fromNamespaceAndPath(Constants.MOD_ID, "fire_creeper_spawns"))
            .add(net.fabricmc.fabric.api.biome.v1.ModificationPhase.ADDITIONS,
                BiomeSelectors.includeByKey(
                    Biomes.NETHER_WASTES,
                    Biomes.SOUL_SAND_VALLEY,
                    Biomes.CRIMSON_FOREST,
                    Biomes.BASALT_DELTAS),
                context -> context.getMobSpawnSettings().addSpawn(
                    MobCategory.MONSTER,
                    new net.minecraft.world.level.biome.MobSpawnSettings.SpawnerData(
                        ModContent.FIRE_CREEPER_ENTITY, 1, 2),
                    25));
        }

        private static void registerEnderSpiderSpawns() {
        BiomeModifications.create(Identifier.fromNamespaceAndPath(Constants.MOD_ID, "ender_spider_spawns"))
            .add(net.fabricmc.fabric.api.biome.v1.ModificationPhase.ADDITIONS,
                BiomeSelectors.includeByKey(
                    Biomes.THE_END,
                    Biomes.END_HIGHLANDS,
                    Biomes.END_MIDLANDS,
                    Biomes.SMALL_END_ISLANDS,
                    Biomes.END_BARRENS),
                context -> context.getMobSpawnSettings().addSpawn(
                    MobCategory.MONSTER,
                    new net.minecraft.world.level.biome.MobSpawnSettings.SpawnerData(
                        ModContent.ENDER_SPIDER_ENTITY, 1, 4),
                    3));
        }

        private static void registerSnowCreeperSpawns() {
        BiomeModifications.create(Identifier.fromNamespaceAndPath(Constants.MOD_ID, "snow_creeper_spawns"))
            .add(net.fabricmc.fabric.api.biome.v1.ModificationPhase.REPLACEMENTS,
                BiomeSelectors.includeByKey(
                    Biomes.SNOWY_PLAINS,
                    Biomes.ICE_SPIKES,
                    Biomes.GROVE,
                    Biomes.SNOWY_SLOPES,
                    Biomes.FROZEN_PEAKS,
                    Biomes.JAGGED_PEAKS,
                    Biomes.SNOWY_BEACH,
                    Biomes.WINDSWEPT_HILLS,
                    Biomes.WINDSWEPT_FOREST,
                    Biomes.WINDSWEPT_GRAVELLY_HILLS,
                    Biomes.STONY_PEAKS,
                    Biomes.TAIGA,
                    Biomes.SNOWY_TAIGA,
                    Biomes.OLD_GROWTH_PINE_TAIGA,
                    Biomes.OLD_GROWTH_SPRUCE_TAIGA),
                context -> {
                    EntityType<?> creeper = BuiltInRegistries.ENTITY_TYPE.get(
                        Identifier.fromNamespaceAndPath("minecraft", "creeper")).orElseThrow().value();
                    context.getMobSpawnSettings().removeSpawnsOfEntityType(creeper);
                    context.getMobSpawnSettings().addSpawn(
                        MobCategory.MONSTER,
                        new net.minecraft.world.level.biome.MobSpawnSettings.SpawnerData(
                            ModContent.SNOW_CREEPER_ENTITY, 4, 4),
                        100);
                });
        }

    private static void registerLootInjections() {
        Map<ResourceKey<LootTable>, ResourceKey<LootTable>> injections = new HashMap<>();
        for (ModLootInjections.Injection injection : ModLootInjections.ABYSSALITE_TEMPLATE_INJECTIONS) {
            injections.put(lootTableKey(injection.targetTable()), lootTableKey(injection.injectedTable()));
        }
        for (ModLootInjections.Injection injection : ModLootInjections.CALAMARI_INJECTIONS) {
            injections.put(lootTableKey(injection.targetTable()), lootTableKey(injection.injectedTable()));
        }
        LootTableEvents.MODIFY.register((key, tableBuilder, source, registries) -> {
            ResourceKey<LootTable> injected = injections.get(key);
            if (injected != null && source.isBuiltin()) {
                tableBuilder.withPool(LootPool.lootPool()
                        .setRolls(ConstantValue.exactly(1.0F))
                        .add(NestedLootTable.lootTableReference(injected)));
            }
        });
    }

    private static ResourceKey<LootTable> lootTableKey(String id) {
        return ResourceKey.create(Registries.LOOT_TABLE, Identifier.parse(id));
    }
}

File: neoforge/src/main/kotlin/com/suprememc/SupremeMC.kt
package com.suprememc

import com.suprememc.content.ModContent
import com.suprememc.content.init.ModItems
import com.suprememc.content.init.ModEntities
import com.suprememc.content.init.ModPotions
import com.suprememc.tabs.CreativeTab
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.chat.Component
import net.minecraft.resources.Identifier
import net.minecraft.resources.ResourceKey
import net.minecraft.world.item.CreativeModeTab
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.block.entity.BlockEntityTypes
import net.neoforged.bus.api.IEventBus
import net.neoforged.fml.common.Mod
import net.neoforged.neoforge.data.event.GatherDataEvent
import net.neoforged.neoforge.event.BlockEntityTypeAddBlocksEvent
import net.neoforged.neoforge.event.entity.EntityAttributeCreationEvent
import net.neoforged.neoforge.event.furnace.FurnaceFuelBurnTimeEvent
import net.neoforged.neoforge.event.brewing.RegisterBrewingRecipesEvent
import net.neoforged.neoforge.common.NeoForge
import net.neoforged.neoforge.registries.RegisterEvent

@Mod(Constants.MOD_ID)
class SupremeMC(eventBus: IEventBus) {
    init {
        Constants.LOG.info("Hello NeoForge world!")
        eventBus.addListener(SupremeMC::registerContent)
        eventBus.addListener(SupremeMC::addSignBlockEntityBlocks)
        eventBus.addListener(SupremeMC::registerAttributes)
        NeoForge.EVENT_BUS.addListener(SupremeMC::registerBrewingRecipes)
            NeoForge.EVENT_BUS.addListener(SupremeMC::registerFuelValues)
        eventBus.addListener(SupremeMC::gatherData)
    }

    companion object {
        // Vanilla's sign block entities only accept their hardcoded block list, so opt the palm signs in.
        private fun addSignBlockEntityBlocks(event: BlockEntityTypeAddBlocksEvent) {
            event.modify(BlockEntityTypes.SIGN, ModContent.PALM_SIGN, ModContent.PALM_WALL_SIGN)
            event.modify(BlockEntityTypes.HANGING_SIGN, ModContent.PALM_HANGING_SIGN, ModContent.PALM_WALL_HANGING_SIGN)
            event.modify(BlockEntityTypes.FURNACE, ModContent.BLACKSTONE_FURNACE, ModContent.DEEPSLATE_FURNACE)
        }
        private fun registerAttributes(event: EntityAttributeCreationEvent) {
            event.put(ModEntities.GRIZZLY_BEAR_ENTITY, net.minecraft.world.entity.animal.polarbear.PolarBear.createAttributes().build())
            event.put(ModEntities.ENDER_SPIDER_ENTITY, com.suprememc.content.entity.EnderSpider.createAttributes().build())
            event.put(ModEntities.FIRE_CREEPER_ENTITY, net.minecraft.world.entity.monster.Creeper.createAttributes().build())
            event.put(ModEntities.SNOW_CREEPER_ENTITY, net.minecraft.world.entity.monster.Creeper.createAttributes().build())
        }
        private fun registerFuelValues(event: FurnaceFuelBurnTimeEvent) {
            when (event.itemStack.item) {
                ModItems.ANTHRACITE -> event.setBurnTime(1600)
                ModContent.ANTHRACITE_BLOCK.asItem() -> event.setBurnTime(16000)
            }
        }
        private fun registerBrewingRecipes(event: RegisterBrewingRecipesEvent) {
            val builder = event.builder
            builder.addMix(net.minecraft.world.item.alchemy.Potions.AWKWARD, ModContent.CLOVER.asItem(), ModPotions.LUCK_POTION)
            builder.addMix(ModPotions.LUCK_POTION, net.minecraft.world.item.Items.REDSTONE, ModPotions.LONG_LUCK_POTION)
            builder.addMix(ModPotions.LUCK_POTION, net.minecraft.world.item.Items.GLOWSTONE_DUST, ModPotions.STRONG_LUCK_POTION)
            builder.addMix(ModPotions.LUCK_POTION, net.minecraft.world.item.Items.FERMENTED_SPIDER_EYE, ModPotions.BAD_LUCK_POTION)
            builder.addMix(ModPotions.LONG_LUCK_POTION, net.minecraft.world.item.Items.FERMENTED_SPIDER_EYE, ModPotions.LONG_BAD_LUCK_POTION)
            builder.addMix(ModPotions.STRONG_LUCK_POTION, net.minecraft.world.item.Items.FERMENTED_SPIDER_EYE, ModPotions.STRONG_BAD_LUCK_POTION)
            builder.addMix(net.minecraft.world.item.alchemy.Potions.AWKWARD, net.minecraft.world.item.Items.ROTTEN_FLESH, ModPotions.HUNGER_POTION)
            builder.addMix(ModPotions.HUNGER_POTION, net.minecraft.world.item.Items.REDSTONE, ModPotions.LONG_HUNGER_POTION)
            builder.addMix(ModPotions.HUNGER_POTION, net.minecraft.world.item.Items.GLOWSTONE_DUST, ModPotions.STRONG_HUNGER_POTION)
            builder.addMix(net.minecraft.world.item.alchemy.Potions.AWKWARD, net.minecraft.world.item.Items.WITHER_ROSE, ModPotions.DECAY_POTION)
            builder.addMix(ModPotions.DECAY_POTION, net.minecraft.world.item.Items.REDSTONE, ModPotions.LONG_DECAY_POTION)
            builder.addMix(ModPotions.DECAY_POTION, net.minecraft.world.item.Items.GLOWSTONE_DUST, ModPotions.STRONG_DECAY_POTION)
            builder.addContainerRecipe(net.minecraft.world.item.Items.SPLASH_POTION, ModItems.EXPERIENCE_DUST, net.minecraft.world.item.Items.SPLASH_POTION)
        }
        private fun registerContent(event: RegisterEvent) {
            if (event.registryKey == Registries.BLOCK || event.registryKey == Registries.ITEM) {
                ModContent.bootstrap()
            } else if (event.registryKey == Registries.CREATIVE_MODE_TAB) {
                val tab = CreativeModeTab.builder(CreativeModeTab.Row.TOP, 0)
                    .title(Component.translatable("itemGroup.${Constants.MOD_ID}.main"))
                    .icon { ItemStack(ModContent.SUPREME_MC_LOGO_BLOCK) }
                    .displayItems { params, output ->
                        ModContent.CREATIVE_TAB_ITEMS.forEach(output::accept)
                        CreativeTab.enchantedBooks(params.holders()).forEach(output::accept)
                        CreativeTab.potions().forEach(output::accept)
                    }
                    .build()
                Registry.register(
                    BuiltInRegistries.CREATIVE_MODE_TAB,
                    ResourceKey.create(
                        Registries.CREATIVE_MODE_TAB,
                        Identifier.fromNamespaceAndPath(Constants.MOD_ID, "main")
                    ),
                    tab
                )
            }
        }

        private fun gatherData(event: GatherDataEvent.Client) {
            val output = event.generator.packOutput
            event.addProvider(AquamarineDataProvider(output))
            event.addProvider(BurningDiamondDataProvider(output))
            event.addProvider(AmberDataProvider(output))
            event.addProvider(AnthraciteDataProvider(output))
            event.addProvider(XyliumDataProvider(output))
            event.addProvider(DustDataProvider(output))
            event.addProvider(PrismarineDataProvider(output))
            event.addProvider(MaterialBlocksDataProvider(output))
            event.addProvider(SupremeMCLogoBlockDataProvider(output))
            event.addProvider(PolishedStoneWallsDataProvider(output))
            event.addProvider(StoneBrickDataProvider(output))
            event.addProvider(EmeraldDataProvider(output))
            event.addProvider(ExperienceDataProvider(output))
            event.addProvider(AbyssaliteDataProvider(output))
            event.addProvider(IcicleDataProvider(output))
            event.addProvider(PalmDataProvider(output))
            event.addProvider(LavenderDataProvider(output))
            event.addProvider(CottonDataProvider(output))
            event.addProvider(CornDataProvider(output))
            event.addProvider(GrapeVineDataProvider(output))
            event.addProvider(TomatoDataProvider(output))
            event.addProvider(ButtercupCloverDataProvider(output))
            event.addProvider(BeachGrassDataProvider(output))
            event.addProvider(CalamariDataProvider(output))
                event.addProvider(DrownedDataProvider(output))
               event.addProvider(EnchantmentDataProvider(output))
            event.addProvider(SmeltingDataProvider(output))
            event.addProvider(FloridaPlainsDataProvider(output))
            event.addProvider(CaysDataProvider(output))
            event.addProvider(IceCavesDataProvider(output))
            event.addProvider(WoodCuttingDataProvider(output))
			event.addProvider(NetherBoatDataProvider(output))
            event.addProvider(GlowSlimeDataProvider(output))
            event.addProvider(GrizzlyBearDataProvider(output))
            event.addProvider(EnderSpiderDataProvider(output))
            event.addProvider(FireCreeperDataProvider(output))
            event.addProvider(SnowCreeperDataProvider(output))
            event.addProvider(SnowTntDataProvider(output))
            event.addProvider(FireTntDataProvider(output))
			event.addProvider(GlendstoneDataProvider(output))
            event.addProvider(BellDataProvider(output))
            event.addProvider(CakeDataProvider(output))
            event.addProvider(MilkDataProvider(output))
            event.addProvider(BookshelfDataProvider(output))
            event.addProvider(CraftingTableDataProvider(output))
            event.addProvider(FurnaceDataProvider(output))
            event.addProvider(MineralTagsDataProvider(output))
            event.addProvider(ColoredSandstoneDataProvider(output))
            event.addProvider(SupremeMCLanguageProvider(output))
        }
    }
}